samza-core/src/main/java/org/apache/samza/clustermanager/ContainerManager.java

```diff
@@ -161,4 +237,140 @@ void handleExpiredRequestWithHostAffinityEnabled(String processorId, String pref
+    return actionStatus;
+  }
+
+  public Optional<Long> getActionExpiryTimeout(String processorId) {
+    return this.actions.containsKey(processorId) ? Optional.of(
+        this.actions.get(processorId).getRequestActionExpiryTimeout()) : Optional.empty();
+  }
+
+  /**
+   * Handles the container allocation for an existing container placement action by issuing a stop on the active container and
+   * waiting for the active container to shutdown for a timeout of {@code actionMetaData#getRequestActionExpiryTimeout}.
+   *
+   * Case 1. If active container fails to stop mark the container placement action failed.
+   * Case 2. Otherwise once active container shuts down then issue a start for the container on the preferred host.
+   *
+   * This method is invoked by the allocator thread which waits on notify from the thread issuing
+   * callbacks i.e {@link ContainerProcessManager}
+   */
+  private void handleContainerAllocationForExistingControlAction(String processorId, ContainerAllocator allocator,
+      ResourceRequestState state, SamzaResourceRequest request, String preferredhost) {
+    // check if container is already dead without issuing a stop here, fail the move request
+    ContainerPlacementMetadata actionMetaData = getControlActionMetadata(processorId).get();
+    if (samzaApplicationState.runningProcessors.containsKey(processorId)) {
+      LOG.info("Requesting running container to shutdown due to existing container placement action {}",
+          getControlActionMetadata(processorId).get());
+      clusterResourceManager.stopStreamProcessor(samzaApplicationState.runningProcessors.get(processorId));
+      Object activeContainerStopped = actionMetaData.getActiveContainerStopped();
+      try {
+        synchronized (activeContainerStopped) {
+          activeContainerStopped.wait(actionMetaData.getRequestActionExpiryTimeout());
+        }
+        LOG.info("Active container is stopped issuing a run on the allocated resource");
+        allocator.runStreamProcessor(request, preferredhost);
+      } catch (InterruptedException e) {
+        e.printStackTrace();
+        state.cancelResourceRequest(request);
+        handleControlActionFailure(processorId, actionMetaData,
+            "timed out waiting for active container to stop with a timeout (ms) of: "
+                + actionMetaData.getRequestActionExpiryTimeout());
+      }
+    }
+  }
+
+  private void handleControlActionFailure(String processorId, ContainerPlacementMetadata metaData, String failureMessage) {
+    metaData.setActionStatus(ContainerPlacementStatus.StatusCode.FAILED, failureMessage);
+    LOG.info("Control Action failed with metadata {}", metaData);
+    this.actions.remove(processorId);
+  }
+
+  /**
+   * Check if a activeContainerResource has control-action-metadata associated with it
+   */
+  private Optional<ContainerPlacementMetadata> getControlActionMetadata(String processorId) {
```

Comment:
I'd suggest not wrapping this up in optional and instead have a method to determine if a processorId is associated with a control action and use that check everywhere this method is used and directly fetch the `ContainerPlacementMetadata`. 

IMO, the fact this is wrapped in `Optional` and we still end up doing `.isPresent`, `.get` and not have any functional programming usage defeats the brevity and the point of having optional.